example/lib/chat_bubble.dart

```diff
@@ -42,58 +49,137 @@ class ChatBubble extends StatelessWidget {
+              .asUint8List());
+    }
+    if (widget.index == null && widget.path == null && file?.path == null) {
+      return;
+    }
+    // Prepare player with extracting waveform if index is even.
+    controller.preparePlayer(
+      path: widget.path ?? file!.path,
+      shouldExtractWaveform: widget.index?.isEven ?? true,
+    );
+    // Extracting waveform separately if index is odd.
+    if (widget.index?.isOdd ?? false) {
+      controller.extractWaveformData(
+        path: widget.path ?? file!.path,
+        noOfSamples: playerWaveStyle.getSamplesForWidth(widget.width ?? 200),
+      );
+    }
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    playerStateSubscription.cancel();
+    if (widget.isLast) {
+      controller.stopAllPlayers();
+    }
+    controller.dispose();
+  }
+
   @override
   Widget build(BuildContext context) {
-    return Align(
-      alignment: isSender ? Alignment.centerRight : Alignment.centerLeft,
-      child: Container(
-        padding: EdgeInsets.only(
-          bottom: 6,
-          right: isSender ? 0 : 10,
-          top: 6,
-        ),
-        margin: const EdgeInsets.symmetric(vertical: 8, horizontal: 12),
-        decoration: BoxDecoration(
-          borderRadius: BorderRadius.circular(10),
-          color: isSender ? const Color(0xFF276bfd) : const Color(0xFF343145),
-        ),
-        child: Row(
-          mainAxisSize: MainAxisSize.min,
-          children: [
-            IconButton(
-              onPressed: onTap,
-              icon: Icon(isPlaying ? Icons.stop : Icons.play_arrow),
-              color: Colors.white,
-              splashColor: Colors.transparent,
-              highlightColor: Colors.transparent,
-            ),
-            AudioFileWaveforms(
-              size: Size(MediaQuery.of(context).size.width / 2, 70),
-              playerController: playerController,
-              playerWaveStyle: const PlayerWaveStyle(
-                fixedWaveColor: Colors.white54,
-                liveWaveColor: Colors.blueAccent,
-                spacing: 6,
+    return widget.path != null || file?.path != null
+        ? Align(
+            alignment:
+                widget.isSender ? Alignment.centerRight : Alignment.centerLeft,
+            child: Container(
+              padding: EdgeInsets.only(
+                bottom: 6,
+                right: widget.isSender ? 0 : 10,
+                top: 6,
+              ),
+              margin: const EdgeInsets.symmetric(vertical: 8, horizontal: 12),
+              decoration: BoxDecoration(
+                borderRadius: BorderRadius.circular(10),
+                color: widget.isSender
+                    ? const Color(0xFF276bfd)
+                    : const Color(0xFF343145),
+              ),
+              child: Row(
+                mainAxisSize: MainAxisSize.min,
+                children: [
+                  if (controller.playerState != PlayerState.stopped)
+                    IconButton(
```

Comment:
If these states come from our plugin than please make extensions for same like isStopped